### PR TITLE
 [IMP] proc_auto_create_grp_by_product: add lock

### PR DESCRIPTION
**Pull request by TDu**

### CODE REVIEW

File: procurement_auto_create_group_by_product/models/stock_rule.py [hunk 1]
```diff
@@ -13,6 +13,10 @@ def _get_auto_procurement_group(self, product):
         if self.auto_create_group_by_product:
             if product.auto_create_procurement_group_ids:
                 return fields.first(product.auto_create_procurement_group_ids)
+            else:
+                # Make sure that two transactions can not create a procurement group
+                # For the same product at the same time.
+                self.env.cr.execute("LOCK TABLE procurement_group NOWAIT;")
```
Comment (mt-software-de) [inline procurement_auto_create_group_by_product/models/stock_rule.py]: I think we shouldn't block the whole table. 
```suggestion
            # Make sure that two transactions can not create a procurement group
            # For the same product at the same time.
            import hashlib
            import struct
            lock_name = f"product.product,{product.id}-auto-proc-group"
            hasher = hashlib.sha1(str(lock_name).encode())
            int_lock = struct.unpack("q", hasher.digest()[:8])
            self.env.cr.execute("SELECT pg_try_advisory_xact_lock(%s);", (int_lock,))
            lock_acquired = self.env.cr.fetchone()[0]
            if not lock_acquired:
                raise UserError(f"The auto procurement group for product {product.name} is already created by someone else")
```

**Review by mt-software-de — CHANGES_REQUESTED** [review]

File: procurement_auto_create_group_by_product/models/stock_rule.py (hunk 1 above)
Comment (TDu) [reply procurement_auto_create_group_by_product/models/stock_rule.py]: Never used advisory lock. Nice thanks for that.
I also added a test for this change.

**Review by mt-software-de — APPROVED** [review]

**Review by sebalix — APPROVED** [review]

File: procurement_auto_create_group_by_product/models/stock_rule.py [hunk 2]
```diff
@@ -13,6 +17,21 @@ def _get_auto_procurement_group(self, product):
...
+                # For the same product at the same time.
+                lock_name = f"product.product,{product.id}-auto-proc-group"
+                hasher = hashlib.sha1(str(lock_name).encode())
+                int_lock = struct.unpack("q", hasher.digest()[:8])
+                self.env.cr.execute(
+                    "SELECT pg_try_advisory_xact_lock(%s);", (int_lock,)
+                )
+                lock_acquired = self.env.cr.fetchone()[0]
+                if not lock_acquired:
+                    raise UserError(
```
Comment (jbaudoux) [inline procurement_auto_create_group_by_product/models/stock_rule.py:31]: Shoudn't we raise a `RetryableJobError` instead ?

Comment (sebalix) [reply procurement_auto_create_group_by_product/models/stock_rule.py:31]: The current module and its parent do not depend on `queue_job`, we would need a glue module for that...

Comment (jbaudoux) [reply procurement_auto_create_group_by_product/models/stock_rule.py:31]: Indeed. Then we should raise something like
```
raise OperationalError(
  "Failed to acquire advisory lock",
  params=None,
  orig=RuntimeError(psycopg2.errorcodes.LOCK_NOT_AVAILABLE)
)
```
which will be properly catched by job runner or odoo service depending on who controls the transaction

**Pull request merged**